### ISSUE TITLE
Byte search syntax

### ISSUE THREAD

**Issue by mbarbar** [krpors/hx#13]
I wouldn't mind giving non-ASCII byte searching a shot. Though I'm not sure what the intented syntax is. editor.c seems to imply any search string starting with "0x" should be regarded as a hex byte sequence though of course this means searching for "0x" is impossible.

Here are the options I think could be possible... the existing input buffer size is 80 and I might make reference to that.

**A separate search command**
That is, instead of /, some other key.
Advantages
- Straightforward to implement (convert typed bytes into actual bytes and use existing search process)
- No extra typing (i.e. no need for \x or 0x) so 40 bytes fit
- No need for a character to become an escape character

Disadvantages
- Not possible to mix bytes and ASCII without typing the hex ASCII value ((e.g. 41 instead of A)
- Creates a separate command

**\x regards the entire search stirng as byte values**
So "/\xabcf" searches for 0xab follow by 0xcf.
Advantages
- Uses same command
- No extra typing past initial "\x", so 39 bytes fit

Disadvantages
- Not possible to mix bytes and ASCII without typing the hex ASCII value
- \ becomes an escape character - though if \ appears in a position after the first, it's not an issue.

**\x regards the following two bytes as a byte value**
So "/a\x41b" searches for a followed by 0x41 followed by b.
Advantages
- Can mix ASCII and byte values
- Uses same command
- Pretty simple to explain in documentation

Disadvantages
- \x needs to be typed before every byte value so a maximum of 20 byte values can be searched (or 19 byte values and 4 ASCII characters, and so on)
- \ becomes an escape character anywhere in the string - \\ is required to search for \ and ' followed by anything but x or \ is an error that needs to be handled

\x can be replaced by 0x though I think making 0 an escape character would be frustrating as it isn't standard (as far as I've seen at least).

**Comment by krpors** [comment]
Searching for a byte sequence was more or less an idea, so I didn't give it much thought :) Anyway, you had me thinking. First of all, the input buffer size was arbitrarily chosen, but 80 was a bit on the safe side because I guess most people use minimum of 80 chars for terminal width (totally not backed by proof...).

My intuition says that a using `\x` as an escape sequence (where you can mix plain ASCII and hex) is the better choice (your last suggestion), for ex.

    /blah\x00null separated values\x00yadda

I'm not entirely sure yet though how this will work out. Are there any other (tty-like) hex editors which provide this functionality? I'm wondering how (and IF) they provide this. 

Edit: and using `0x` as an escape sequence is just horrible :D

**Comment by mbarbar** [comment]
I think the 80 minimum is a fair assumption. After all, lots of style guides indicate max line length of 79/80.

>I'm not entirely sure yet though how this will work out. Are there any other (tty-like) hex editors which provide this functionality? I'm wondering how (and IF) they provide this.

I had a very quick look at bless and dhex. Unless I missed something they allow you only to search on separate types, either ASCII **or** bytes. dhex is interesting though, you can input your ASCII search string into the prompt then switch to the byte search prompt, and it'll convert all what you entered into the equivalent hex values.

I also checked Vim and you can mix them. "\%x41b" searches for Ab (oddly enough, since I have case insensitivity enabled for searches, it searches for ab/Ab/aB/AB despite the fact that 41 represents upper case a, but if I search for "\%x41B" it will only search for AB).

So in conclusion, from my not-so-comprehensive research, ASCII-or-byte is popular, and Vim, being the editor that provides everything, allows for mixing in a way similar to the last suggestion.

Implementation wise, I think the challenge is just parsing and clean error reporting; the rest is simple because search is implemented.

Edit: w.r.t your example, in Vim that'd be almost identical - /blah\%x00null separated values\%x00yadda
